### DISCUSSION TITLE
Making Randomizable transform threadsafe by having randomize() return the required random value

### DISCUSSION THREAD

**Discussion by RomStriker** [Project-MONAI/MONAI#6951] [Q&A] [answered]
Hi,

For context, I am using CacheDataset with ThreadDataloader. I have a random transform class `CustomTransform(Randomizable, Transform)` which inherits from both `Randomizable` and `Transform`. In the documentation for [Randomizable](https://docs.monai.io/en/latest/transforms.html#randomizable) it says "This API is not thread-safe.". 

I noticed that the abstract `randomize()` method of `Randomizable` does not return anything and in examples it is used only to update class attribute randomly, which can then be used further. This I assume could be a problem when used with a ThreadDataloader, since a different thread might update the random class attribute before the main thread can use the previous value of the class attribute. 

I thought I could solve this by making the randomize() method return the random value instead of storing it in a class attribute and using it directly. Would this approach make this Randomizable transform threadsafe? Maybe there is some aspect of randomness that I am missing. 

Also, is there some simple way of making Randomizable methods threadsafe? In the documentation for [Transforms](https://docs.monai.io/en/latest/transforms.html#transform), it mentions "When used from a multi-process context, transform’s instance variables are read-only. thread-unsafe transforms should inherit `monai.transforms.ThreadUnsafe`." What does this mean? 

Any comments would be appreciated.

**Comment by wyli** [comment]
thanks for looking into this, in some use cases we want to generate the randomised factor once and apply it across different modalities/image-label pairs consistently, the class attributes are used to synchronise these..

**Reply by RomStriker** [reply]
Okay, thanks. Do you think returning the randomised factor if I only need to use it once makes it thread-safe?

**Reply by wyli** [reply, marked as answer]
Yes, I think that'll be thread-safe.
